### PR TITLE
Update model to include tags and add assign command

### LINKED ISSUE #71: Assign vendor to an event
Implement the `assign` command as per defined in the feature specification

Should also update event model to contain the vendors assigned
updating vendor model to contain the event assigned to is optional(?)

### PR DESCRIPTION
**Pull request by kevin-pek**
This PR is be a bit larger in scope since it is the first of the 2 to be made for the assign commands. Resolves #71.

Changes:
- `AssignCommand` and `AssignCommandParser` implementation.
- Update `AddressBook`, `Model` and `ModelManager` to support maintaining a list of unique associations between vendors and events.

>Requires `TypicalEvents` class from #87 to be merged before this PR can pass CI.

### CODE REVIEW

File: src/main/java/seedu/address/model/AddressBook.java [hunk 1]
```diff
@@ -20,6 +24,7 @@ public class AddressBook implements ReadOnlyAddressBook {
 
     private final UniqueVendorList vendors;
     private final UniqueEventList events;
+    private final Set<Pair<Vendor, Event>> associations;
```
Comment (C5hives) [inline src/main/java/seedu/address/model/AddressBook.java:27]: You can consider abstracting away this data structure into its own class similar to what was done for `UniqueVendorList` and `UniqueEventList`.

Perhaps a `VendorEventAssociations` class would be better?

File: src/main/java/seedu/address/model/AddressBook.java [hunk 2]
```diff
@@ -108,6 +114,28 @@ public void removeVendor(Vendor key) {
...
+    //// assign operations
+
+    /**
+     * Returns true if the given {@code vendor} is already assigned to the given {@code event}.
+     * {@code vendor} and {@code event} must exist in the address book.
+     */
+    public boolean isVendorAssignedToEvent(Vendor vendor, Event event) {
+        requireAllNonNull(vendor, event);
+        Pair<Vendor, Event> pair = new Pair<>(vendor, event);
+        return associations.contains(pair);
```
Comment (C5hives) [inline src/main/java/seedu/address/model/AddressBook.java:126]: These lines requires `AddressBook` class to know about the implementation details of the `associations` variable.

As metioned in the previous comment, it might be better to abstract such details away

File: src/test/java/seedu/address/logic/parser/AssignCommandParserTest.java
Comment (C5hives) [inline src/test/java/seedu/address/logic/parser/AssignCommandParserTest.java:1]: Tests are very comprehensive—nice!

Would it be possible to merge the test methods that test similar invalid syntax (e.g. missing vendor prefix, missing event prefix) into a single method?
The current `AddCommandParserTest` does this (see test method `parse_repeatedNonTagValue_failure()`)

File: src/main/java/seedu/address/logic/parser/AssignCommandParser.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AssignCommand.MESSAGE_USAGE));
+        }
+
+        // Parse vendor index using the Index class
+        Index vendorIndex;
+        try {
+            vendorIndex = ParserUtil.parseIndex(argMultimap.getValue(PREFIX_VENDOR).get());
+        } catch (ParseException pe) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AssignCommand.MESSAGE_USAGE), pe);
+        }
```
Comment (C5hives) [inline src/main/java/seedu/address/logic/parser/AssignCommandParser.java:36]: Is this try-catch block (and the subsequent one) needed?
- `ParserUtil` already throws `ParserException` when the parameter values do not adhere to their constraints (such as a `Name` having special characters).
- By catching the `ParserException` and throwing a new one with `MESSAGE_USAGE`, the user might not fully understand why the command failed to be parsed.

I'd argue that an error with formatting is very different from an error due to invalid values entered by the user. Would it be clearer for the user if the `ParseException` from `ParserUtil` is displayed instead?

File: src/main/java/seedu/address/logic/commands/AssignCommand.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        if (eventIndex.getZeroBased() >= eventList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_EVENT_DISPLAYED_INDEX);
+        }
+
+        Vendor vendor = vendorList.get(vendorIndex.getZeroBased());
+        Event event = eventList.get(eventIndex.getZeroBased());
+
+        if (model.isVendorAssignedToEvent(vendor, event)) {
+            throw new CommandException(Messages.MESSAGE_VENDOR_ALREADY_ASSIGNED);
+        }
```
Comment (C5hives) [inline src/main/java/seedu/address/logic/commands/AssignCommand.java:76]: Nice guard clauses to systemically rule out invalid states—the different error messages are also helpful for diagnosing issues with the command execution :D

File: src/main/java/seedu/address/logic/commands/AssignCommand.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+        model.assignVendorToEvent(vendor, event);
+
+        return new CommandResult(
+                String.format(MESSAGE_ASSIGN_SUCCESS,
+                vendorIndex.getOneBased(),
+                eventIndex.getOneBased()));
```
Comment (C5hives) [inline src/main/java/seedu/address/logic/commands/AssignCommand.java:83]: Is the indentation for lines 82 and 83 wrong? I might be mistaken here

File: src/main/java/seedu/address/model/AddressBook.java [hunk 6]
```diff
@@ -108,6 +114,28 @@ public void removeVendor(Vendor key) {
         vendors.remove(key);
     }
 
+    //// assign operations
+
+    /**
+     * Returns true if the given {@code vendor} is already assigned to the given {@code event}.
+     * {@code vendor} and {@code event} must exist in the address book.
+     */
+    public boolean isVendorAssignedToEvent(Vendor vendor, Event event) {
```
Comment (C5hives) [inline src/main/java/seedu/address/model/AddressBook.java:123]: Good method name that explicitly describes the relationship between a `Vendor` and an `Event`!

File: src/main/java/seedu/address/model/Model.java [hunk 7]
```diff
@@ -93,6 +93,17 @@ public interface Model {
...
 
+    /**
+     * Returns true if the given {@code vendor} is already assigned to the given {@code event}.
+     * {@code vendor} and {@code event} must exist in the address book.
+     */
+    boolean isVendorAssignedToEvent(Vendor vendor, Event event);
+
+    /**
+     * Assigns the vendor corresponding to the id to the event specified by its id.
+     */
```
Comment (C5hives) [inline src/main/java/seedu/address/model/Model.java:104]: Sounds a bit off. Maybe rephrase it to:
- Assigns the given vendor to the given event.

"Given" seems to be the verb of choice for the existing javadocs, so it might make sense to follow it for consistency

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java [hunk 8]
```diff
@@ -54,6 +55,9 @@ public Command parseCommand(String userInput) throws ParseException {
 
         switch (commandWord) {
 
+        case AssignCommand.COMMAND_WORD:
+            return new AssignCommandParser().parse(arguments);
```
Comment (C5hives) [inline src/main/java/seedu/address/logic/parser/AddressBookParser.java:59]: I'd argue that these lines are important enough to not skimp out on their tests. You can consider adding a simple test for this switch case (which shouldn't take too long)

**Review by C5hives — COMMENTED** [review]
Left some suggestions. Some of them are quite subjective so I'll leave it to your discretion ;)

File: src/main/java/seedu/address/logic/parser/AssignCommandParser.java (hunk 3 above)
Comment (C5hives) [reply src/main/java/seedu/address/logic/parser/AssignCommandParser.java:36]: On closer inspection, it seems like the other commands that deal with the `Index` object use this try-catch structure as well.

Since an invalid `Index` is different from other parameters (such as Name) with arbitrary constraints, it would make sense to catch the exception and handle it. Resolving this comment as it is longer valid.

File: src/main/java/seedu/address/logic/commands/AssignCommand.java (hunk 5 above)
Comment (kevin-pek) [reply src/main/java/seedu/address/logic/commands/AssignCommand.java:83]: Nice catch! Not sure why checkstyle didn't flag this out.

**Review by Jasperabez — COMMENTED** [review]
Looks good to me but perhaps need to open a separate issue for getting assigned from vendors and events since @hanscau issue is dependent on that

File: src/main/java/seedu/address/model/AddressBook.java (hunk 1 above)
Comment (kevin-pek) [reply src/main/java/seedu/address/model/AddressBook.java:27]: Will make a new issue to introduce a new class for this!

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java (hunk 8 above)
Comment (Jasperabez) [reply src/main/java/seedu/address/logic/parser/AddressBookParser.java:59]: I agree we need all the test coverage we need sadly..

**Comment by kevin-pek** [comment]
> Looks good to me but perhaps need to open a separate issue for getting assigned from vendors and events since @hanscau issue is dependent on that

Curious, am wondering what else would be needed to support this? My current understanding is that the methods introduced in this PR should suffice 🤔.

File: src/main/java/seedu/address/logic/commands/AssignCommand.java (hunk 4 above)
Comment (C5hives) [reply src/main/java/seedu/address/logic/commands/AssignCommand.java:76]: Resolving as this comment does not require any changes.

File: src/main/java/seedu/address/logic/commands/AssignCommand.java (hunk 5 above)
Comment (C5hives) [reply src/main/java/seedu/address/logic/commands/AssignCommand.java]: Resolving as this issue has been fixed.

File: src/main/java/seedu/address/model/AddressBook.java (hunk 6 above)
Comment (C5hives) [reply src/main/java/seedu/address/model/AddressBook.java:123]: Resolving as this comment does not require any changes.

File: src/main/java/seedu/address/model/Model.java (hunk 7 above)
Comment (C5hives) [reply src/main/java/seedu/address/model/Model.java]: Resolving as this issue has been resolved.

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java (hunk 8 above)
Comment (C5hives) [reply src/main/java/seedu/address/logic/parser/AddressBookParser.java:59]: Resolving as tests for this switch case has been added.

File: src/main/java/seedu/address/model/AddressBook.java (hunk 1 above)
Comment (C5hives) [reply src/main/java/seedu/address/model/AddressBook.java:27]: Resolving as an issue has been created for this.

File: src/main/java/seedu/address/model/AddressBook.java (hunk 2 above)
Comment (C5hives) [reply src/main/java/seedu/address/model/AddressBook.java:126]: Resolving as an issue has been created for this.

**Review by C5hives — APPROVED** [review]
Looks good!
